src/include/btree.i

```diff
@@ -1070,108 +1085,65 @@ __wt_row_leaf_value(WT_PAGE *page, WT_ROW *rip, WT_ITEM *value)
-            *start_ts = 0;
-        if (stop_ts != NULL)
-            *stop_ts = 0;
-        if (start_txn != NULL)
-            *start_txn = 0;
-        if (stop_txn != NULL)
-            *stop_txn = 0;
-    } else if (__wt_off_page(page, addr)) {
-        *addrp = addr->addr;
-        *sizep = addr->size;
-        if (start_ts != NULL)
-            *start_ts = addr->oldest_start_ts;
-        if (start_txn != NULL)
-            *start_txn = addr->oldest_start_txn;
-        if (stop_ts != NULL)
-            *stop_ts = addr->oldest_start_ts;
-        if (stop_txn != NULL)
-            *stop_txn = addr->oldest_start_txn;
-    } else {
-        __wt_cell_unpack(session, page, (WT_CELL *)addr, unpack);
-        *addrp = unpack->data;
-        *sizep = unpack->size;
-        if (start_ts != NULL)
-            *start_ts = unpack->oldest_start_ts;
-        if (start_txn != NULL)
-            *start_txn = unpack->oldest_start_txn;
-        if (stop_ts != NULL)
-            *stop_ts = unpack->newest_stop_ts;
-        if (stop_txn != NULL)
-            *stop_txn = unpack->newest_stop_txn;
-    }
-}
+    WT_ORDERED_READ(addr, ref->addr);
 
-/*
- * __wt_ref_info_lock --
- *     Lock the WT_REF and return the addr/size and type triplet for a reference.
- */
-static inline void
-__wt_ref_info_lock(WT_SESSION_IMPL *session, WT_REF *ref, uint8_t *addr_buf, size_t *sizep)
-{
-    size_t size;
-    uint8_t previous_state;
-    const uint8_t *addr;
+    /* If NULL, there is no information. */
+    if (addr == NULL)
+        return (false);
 
-    /*
-     * The WT_REF address references either an on-page cell or in-memory structure, and eviction
-     * frees both. If our caller is already blocking eviction (either because the WT_REF is locked
-     * or there's a hazard pointer on the page), no locking is required, and the caller should call
-     * the underlying function directly. Otherwise, our caller is not blocking eviction and we lock
-     * here, and copy out the address instead of returning a reference.
-     */
-    for (;; __wt_yield()) {
-        previous_state = ref->state;
-        if (previous_state != WT_REF_LOCKED &&
-          WT_REF_CAS_STATE(session, ref, previous_state, WT_REF_LOCKED))
-            break;
+    /* If off-page, the pointer references a WT_ADDR structure. */
+    if (__wt_off_page(page, addr)) {
+        copy->newest_durable_ts = addr->newest_durable_ts;
+        copy->oldest_start_ts = addr->oldest_start_ts;
+        copy->oldest_start_txn = addr->oldest_start_txn;
+        copy->newest_stop_ts = addr->newest_stop_ts;
+        copy->newest_stop_txn = addr->newest_stop_txn;
+        copy->type = addr->type;
+        memcpy(copy->addr, addr->addr, copy->size = addr->size);
+        return (true);
     }
 
-    __wt_ref_info(session, ref, &addr, &size);
-
-    if (addr_buf != NULL) {
-        if (addr != NULL)
-            memcpy(addr_buf, addr, size);
-        *sizep = size;
+    /* If on-page, the pointer references a cell. */
```

Comment:
Why does this solve the problem of an `addr` getting written to as you're reading it? Couldn't that happen while you're doing this + `memcpy`ing?